lib/api/core/validation/index.js

```diff
@@ -1,690 +1,708 @@
+          .then(validationFilter => {
+            treatedSpecification.validators = validationFilter.id;
+
+            return resolve(treatedSpecification);
+          })
+          .catch(error => {
+            this.kuzzle.pluginsManager.trigger('log:error', error);
+            return reject(new KuzzleInternalError(`Validator specification of the collection ${indexName}.${collectionName} triggered an error`));
+          });
+      }
+      else {
+        return resolve(treatedSpecification);
+      }
+    });
+  }
 
-  return new Promise((resolve, reject) => {
+  structureCollectionValidation(collectionSpec, indexName, collectionName, verboseErrors) {
     var
-      treatedSpecification = {
-        strict: collectionSpec.strict || false,
-        fields: {},
-        validators: null
-      };
+      result,
+      errors = [],
+      fields = {},
+      maxDepth = 0;
 
-    if (!checkAllowedProperties(collectionSpec, ['strict', 'fields', 'validators'])) {
-      errorMessage = `${indexName}.${collectionName}: the collection specification has invalid properties.`;
-      if (verboseErrors) {
-        return resolve({isValid: false, errors: [errorMessage]});
-      }
-      return reject(new InternalError(errorMessage));
-    }
+    verboseErrors = verboseErrors || false;
+
+    Object.keys(collectionSpec.fields).forEach(fieldName => {
+      var
+        field,
+        // We deep clone the field because we will modify it
+        fieldSpecClone = _.cloneDeep(collectionSpec.fields[fieldName]);
 
-    if (collectionSpec.fields) {
       try {
-        result = this.structureCollectionValidation(collectionSpec, indexName, collectionName, verboseErrors);
+        result = this.curateFieldSpecification(fieldSpecClone, indexName, collectionName, fieldName, verboseErrors);
         if (result.isValid === false) {
-          if (verboseErrors) {
-            // do not fail fast if we need verbose errors
-            return resolve(result);
+          errors = _.concat(errors, result.errors);
+          this.kuzzle.pluginsManager.trigger('log:error', result.errors.join('\n'));
+        } else {
+          field = result.fieldSpec;
+          field.path = fieldName.split('/');
+          field.depth = field.path.length;
+          if (field.depth > maxDepth) {
+            maxDepth = field.depth;
+          }
+
+          if (!fields[field.depth]) {
+            fields[field.depth] = [];
           }
-          errorObject = new BadRequestError(result.errors[0]);
-          errorObject.details = result.errors;
-          return reject(errorObject);
+
+          fields[field.depth].push(field);
         }
-        treatedSpecification.fields = result;
       }
       catch (error) {
-        return reject(error);
+        this.kuzzle.pluginsManager.trigger('log:error', error);
+        throwOrStoreError(`Specification for the field ${indexName}.${collectionName}.${fieldName} triggered an error`, verboseErrors, errors);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Split this 143 characters long line (which is greater than 80 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ALineLength)